src/hash_table/bucket_ops.rs

```diff
@@ -0,0 +1,554 @@
+// Copyright 2020 Amari Robinson
+//
+// Licensed under the Apache License, Version 2.0, <LICENSE-APACHE or
+// http://apache.org/licenses/LICENSE-2.0> or the MIT license <LICENSE-MIT or
+// http://opensource.org/licenses/MIT>, at your option. This file may not be
+// copied, modified, or distributed except according to those terms.
+
+use core::marker::PhantomData;
+use core::ptr::NonNull;
+
+use crate::adapter::Adapter;
+use crate::linked_list::{LinkedList, LinkedListOps};
+use crate::pointer_ops::PointerOps;
+use crate::singly_linked_list::{SinglyLinkedList, SinglyLinkedListOps};
+
+/// Trait for bucket operations.
+///
+/// # Time Complexities
+///
+/// |Operation|`SinglyLinkedList`|`LinkedList`|
+/// |-|-|-|
+/// |`is_empty`|`O(1)`|`O(1)`|
+/// |`clear`|`O(n)`|`O(n)`|
+/// |`fast_clear`|`O(1)`|`O(1)`|
+/// |`is_null`|`O(1)`|`O(1)`|
+/// |`get`|`O(1)`|`O(1)`|
+/// |`cursor_from_ptr`|`O(1)`|`O(1)`|
+/// |`find_prev`|`O(n)`|`O(n)`|
```

Comment:
`find_prev` searches the bucket for the a matching element, and returns its predecessor. `BucketOps` lacks a `prev` function because a bucket does not need to be iterated backwards.